### ISSUE TITLE
Huge performance difference with "Prefer 32-bit" on

### ISSUE THREAD

**Issue by George1205** [dotnet/coreclr#8457]
Hi,

According to github.com/dotnet/coreclr/issues/993 , a bug in .Net causes a slow down of certain operations. The item is marked as fixed.

However, after investigating some other code, I have found that either the issue is not fixed (in production), or I have found a similar bug.

Specifically, I have created the attached solution to reproduce the problem. This is consistent on 3 computers, a laptop and 2 desktops, all with Intel processors (2 i5s), with VS 2015 SP1, SP2 and VS 2017 RC. The project targets .Net 4.5 but 4.6.1 is installed on all computers.

The results after running the attached solution is as follows, after two runs for each combination:

1. With "Prefer 32-bit" is off, "useLegacyJit" is not in config, the speed is 26.5 MB/s (12.1 s).

2. With "Prefer 32-bit" is off, "useLegacyJit" is "1", the speed is 69.4 MB/s (4.6 s).

3. With "Prefer 32-bit" is on, "useLegacyJit" is not in config, the speed is 106.5 MB/s (3.0 s). Yes, there is an astonishing 4 times increase.

4. With "Prefer 32-bit" is on, "useLegacyJit" is "1", the speed is 106.5 MB/s (3.0 s).

In the production code, combination 3 is consistently faster than combination 4 with about 2.5 %, but that's clearly irrelevant.

Oh, and one really bizarre thing is that for combination 1, if some private readonly table variables are also declared static, the speed is 21.4 MB/s (14.9 s), so it's worse. For the other combinations, the performance is very slightly better (like 0.5...1 MB / s more).

My problem is that I can't compile the production code with "Prefer 32-bit" on because I have some other code that's optimized to run on 64 bit, code which slows down in a similar way, so there are conflicting needs.

To put things in perspective, the code that needs to run on 32 bit is Rijndael 256 (from which the test presented here was extracted), and the code that needs to run on 64 bit is SHA 512.

Regards,
George
[SpeedTest.zip](https://github.com/dotnet/coreclr/files/630843/SpeedTest.zip)

**Comment by mikedn** [comment]
993 was a very specific issue and affected only floating point conversions. It is certainly fixed and it has nothing to do with your code.

I took a quick look at the generated assembly code and it seems that are multiple CQ issues. Some are specific to 64 bit (the way static fields are accessed, the way arrays indices are sign extended) and some are RyuJIT specific. Some things that caught my eye:

Some JIT32 code:
```asm
   200:                 output[30] = (byte)(SBox[(byte)(b2 >> 8)] ^ (byte)(ExpandedKey[119] >> 8));
00F93D82 8B 45 C8             mov         eax,dword ptr [ebp-38h]  
00F93D85 C1 E8 08             shr         eax,8  
00F93D88 25 FF 00 00 00       and         eax,0FFh  
00F93D8D 8B 15 04 35 DB 03    mov         edx,dword ptr ds:[3DB3504h]  
00F93D93 3B 42 04             cmp         eax,dword ptr [edx+4]  
00F93D96 73 66                jae         00F93DFE  
00F93D98 0F B6 4C 02 08       movzx       ecx,byte ptr [edx+eax+8]  
00F93D9D 83 7E 04 77          cmp         dword ptr [esi+4],77h  
00F93DA1 76 5B                jbe         00F93DFE  
00F93DA3 8B 86 E4 01 00 00    mov         eax,dword ptr [esi+1E4h]  
00F93DA9 C1 E8 08             shr         eax,8  
00F93DAC 25 FF 00 00 00       and         eax,0FFh  
00F93DB1 33 C8                xor         ecx,eax  
00F93DB3 81 E1 FF 00 00 00    and         ecx,0FFh  
00F93DB9 8B 45 08             mov         eax,dword ptr [ebp+8]  
00F93DBC 83 78 04 1E          cmp         dword ptr [eax+4],1Eh  
00F93DC0 76 3C                jbe         00F93DFE  
00F93DC2 88 48 26             mov         byte ptr [eax+26h],cl  
```
Some RyuJIT64 code with comments:
```asm
   200:                 output[30] = (byte)(SBox[(byte)(b2 >> 8)] ^ (byte)(ExpandedKey[119] >> 8));
00007FFB08EEA163 48 BA F0 59 00 90 C2 01 00 00 mov         rdx,1C2900059F0h  
00007FFB08EEA16D 48 8B 12             mov         rdx,qword ptr [rdx]  
; rdx = SBox, looks like it gets spilled and then reloaded. Presumably RyuJIT CSEd the static field
; load and then run out of registers and spilled? But then it reloads in rdx...
00007FFB08EEA170 48 89 54 24 38       mov         qword ptr [rsp+38h],rdx  
00007FFB08EEA175 48 8B 54 24 38       mov         rdx,qword ptr [rsp+38h]  
; Range check. For some reason the length is loaded into a register even though we could have had
; cmp         ecx, dword ptr [rdx+8] below.
00007FFB08EEA17A 8B 52 08             mov         edx,dword ptr [rdx+8]  
00007FFB08EEA17D 41 8B CC             mov         ecx,r12d  
00007FFB08EEA180 C1 E9 08             shr         ecx,8
; Should have been movzx ecx, cl
00007FFB08EEA183 81 E1 FF 00 00 00    and         ecx,0FFh  
00007FFB08EEA189 3B CA                cmp         ecx,edx  
00007FFB08EEA18B 0F 83 E0 00 00 00    jae         00007FFB08EEA271  
00007FFB08EEA191 48 8B 54 24 38       mov         rdx,qword ptr [rsp+38h]  
00007FFB08EEA196 41 8B CC             mov         ecx,r12d
; Now we're doing the shift again. I guess that the range check cloned the index expression...
00007FFB08EEA199 C1 E9 08             shr         ecx,8  
00007FFB08EEA19C 81 E1 FF 00 00 00    and         ecx,0FFh
; Not needed, the original value is a byte so it isn't signed.
00007FFB08EEA1A2 48 63 C9             movsxd      rcx,ecx  
00007FFB08EEA1A5 0F B6 54 0A 10       movzx       edx,byte ptr [rdx+rcx+10h] 
; I have no idea what is this. Looks like `ExpandedKey` was in a register and now it got spilled...
00007FFB08EEA1AA 4C 89 4C 24 30       mov         qword ptr [rsp+30h],r9  
00007FFB08EEA1AF 48 8B 4C 24 30       mov         rcx,qword ptr [rsp+30h]  
; Range check for ExpandedKey
00007FFB08EEA1B4 83 79 08 77          cmp         dword ptr [rcx+8],77h  
00007FFB08EEA1B8 0F 86 B3 00 00 00    jbe         00007FFB08EEA271  
00007FFB08EEA1BE 48 8B 4C 24 30       mov         rcx,qword ptr [rsp+30h]  
00007FFB08EEA1C3 8B 89 EC 01 00 00    mov         ecx,dword ptr [rcx+1ECh]  
00007FFB08EEA1C9 C1 E9 08             shr         ecx,8  
00007FFB08EEA1CC 81 E1 FF 00 00 00    and         ecx,0FFh  
00007FFB08EEA1D2 33 D1                xor         edx,ecx  
00007FFB08EEA1D4 81 E2 FF 00 00 00    and         edx,0FFh
; Spilling again...
00007FFB08EEA1DA 89 94 24 4C 0F 00 00 mov         dword ptr [rsp+0F4Ch],edx  
; Range check for output
00007FFB08EEA1E1 83 78 08 1E          cmp         dword ptr [rax+8],1Eh  
00007FFB08EEA1E5 0F 86 86 00 00 00    jbe         00007FFB08EEA271  
00007FFB08EEA1EB 8B 94 24 4C 0F 00 00 mov         edx,dword ptr [rsp+0F4Ch]  
00007FFB08EEA1F2 88 50 2E             mov         byte ptr [rax+2Eh],dl  
```

As it can be seen above the x64 code is much larger and that's partly due to the fact that the JIT can't keep values in registers.

cc @dotnet/jit-contrib

**Comment by George1205** [comment]
Thank you for looking into this.

I don't know if this is relevant, but you've mentioned range checking. I have also tested the code (all those assignments) by putting it within "unchecked", but it makes no difference (performance wise) whether in 32 or 64 bit.

**Comment by mikedn** [comment]
`unchecked` doesn't affect array range checks, currently the only way to avoid those is to use `unsafe` and pointers (BTW I wanted to suggest that as a possible workaround but forgot).

Anyway, not that the JIT32 version has range checks as well and works pretty well. In my experience range checks aren't as bad (performance wise) as some may think. For the sake of completeness I'll update the post to show range checks in the JIT32 code.

**Comment by mikedn** [comment]
> ; rdx = SBox, looks like it gets spilled and then reloaded. Presumably RyuJIT CSEd the static field
; load and then run out of registers and spilled? But then it reloads in rdx...

This is actually caused by `fgMorphArrayIndex` which has to ensure that the range check and the array access use the same array and index values so it puts them in temporaries.

> ; Now we're doing the shift again. I guess that the range check cloned the index expression...

And the duplicate shift is also generated by `fgMorphArrayIndex` which prefers to clone simple expressions instead of introducing new temporaries. I wonder if the range check can't be made to return the checked index so there's no need for such duplication (nor additional temps).

> ; I have no idea what is this. Looks like `ExpandedKey` was in a register and now it got spilled...

It looks like this is caused by `fgMorphArrayIndex` and CSE, the first introduces a temp and then CSE introduces a another temp on top of the existing temp.

> ; Spilling again...

That's another temp introduced by the importer for "Strict ordering of exceptions for Array store".

In summary: we get a lot of temporaries early and later phases don't handle that too well.

**Comment by RussKeldorph** [comment]
@George1205 since you are using the .NET Framework, you may want to report this via [Connect](http://connect.microsoft.com).  It's relevant to .NET Core, however, so I'll leave it open here.

**Comment by George1205** [comment]
I've added the issue on Connect as well, but I wasn't able to attach the zip file: https://connect.microsoft.com/VisualStudio/feedback/details/3115697/huge-performance-difference-with-prefer-32-bit

**Comment by roterdam** [comment]
@mikedn why can't the 64-bit jit keep the values in the registers?

**Comment by mikedn** [comment]
> why can't the 64-bit jit keep the values in the registers?

There's a limit to the number of local variables the JIT tracks during data flow analysis - 512. Variables are sorted by their reference count and those that exceed the limit aren't tracked and end up on the stack.

That said, JIT32 has a lower limit - 128. So it's possible that the limit isn't itself a problem, the large number of local variables is. The IL code for this method has 20 variables but the time we reach register allocation we have almost 650, every array access (and there are lots of them in this code) introduces new local variables.

Increasing the limit to 1024 does generate better code:
```asm
       48B81828C76BD9010000 mov      rax, 0x1D96BC72818
       488B00               mov      rax, gword ptr [rax]
       8BD1                 mov      edx, ecx
       C1EA08               shr      edx, 8
       0FB6D2               movzx    rdx, dl
       3B5008               cmp      edx, dword ptr [rax+8]
       0F837D000000         jae      G_M59478_IG103
; this and the following movzx are duplicate of the above
       C1E908               shr      ecx, 8
       0FB6D1               movzx    rdx, cl
 ; this is redundant (though perhaps more difficult to eliminate)
       4863D2               movsxd   rdx, edx
       0FB6441010           movzx    rax, byte  ptr [rax+rdx+16]
       488BD7               mov      rdx, rdi
       83FB77               cmp      ebx, 119
       7667                 jbe      SHORT G_M59478_IG103
       8B92EC010000         mov      edx, dword ptr [rdx+492]
       C1EA08               shr      edx, 8
 ; this is useless due to the movzx below
       0FB6D2               movzx    rdx, dl
       33C2                 xor      eax, edx
 ; this is useless due to the byte store that uses it
       0FB6C0               movzx    rax, al
       837D081E             cmp      dword ptr [rbp+8], 30
       7650                 jbe      SHORT G_M59478_IG103
       88452E               mov      byte  ptr [rbp+46], al
```
But this still only does 60MB/sec instead of 100. Maybe it's because I'm running a checked CoreCLR build but I don't see how that could affect this particular code. It's possible that the remaining issues (some redundant casts and shifts) are also important.

**Comment by George1205** [comment]
I've rerun some tests with just a single round (12). I wanted to have as little code as possible, to hopefully avoid the effect mentioned by Mike about the local variables.

(I've increased the number of iterations, "count", 30 times.)

The results are ("useLegacyJit" is not in config):

1. With "Prefer 32-bit" off, the speed is 855 MB/s (19.6 for all rounds). The private readonly tables are static.

2. With "Prefer 32-bit" off, the speed is 1980 MB/s (24.5 for all rounds). The private readonly tables are NOT static.

3. With "Prefer 32-bit" on, the speed is 4120 MB/s (45.5 for all rounds). (It's the same with the tables static or not static.)

These tests were done on the laptop, so the performance is much lower than for the desktops, and the ratios are also different.

Looking at the ratios, the performance is much worse for one round than for all the rounds.

**Comment by mikedn** [comment]
> I've rerun some tests with just a single round (12). I wanted to have as little code as possible, to hopefully avoid the effect mentioned by Mike about the local variables.

It would seem that in .NET Framework you still end up with quite a few variables, this can be seen in the function prolog:
```asm
00007FF8DA2307FC 48 81 EC 18 02 00 00 sub         rsp,218h  
00007FF8DA230803 48 8B F1             mov         rsi,rcx  
00007FF8DA230806 48 8D 7C 24 20       lea         rdi,[rsp+20h]  
00007FF8DA23080B B9 5C 00 00 00       mov         ecx,5Ch
00007FF8DA230810 33 C0                xor         eax,eax
00007FF8DA230812 F3 AB                rep stos    dword ptr [rdi]  ; zero out 368 bytes of local vars
00007FF8DA230814 48 8B CE             mov         rcx,rsi  
```
CoreCLR doesn't do this so I can't investigate what's going on, the current .NET Framework release is quite a bit behind CoreCLR.

BTW, the original code does better in .NET Core than in .NET Framework - 50 MB/sec vs. 35 MB/sec.

It's also interesting to see the function prolog for the original code, I should have looked at that before:

.NET Framework zeroes out 3880 bytes:
```asm
00007FF8DA2107FC 48 81 EC D8 0F 00 00 sub         rsp,0FD8h  
00007FF8DA210803 48 8B F1             mov         rsi,rcx  
00007FF8DA210806 48 8D 7C 24 20       lea         rdi,[rsp+20h]  
00007FF8DA21080B B9 CA 03 00 00       mov         ecx,3CAh
00007FF8DA210810 33 C0                xor         eax,eax  
00007FF8DA210812 F3 AB                rep stos    dword ptr [rdi]  
```
.NET Core zeroes out 824 bytes. That's a good improvement but it's still a lot considering that from the developer's perspective this code writes only 32 bytes to memory.
```asm
       4881EC18050000       sub      rsp, 0x518
       488BF1               mov      rsi, rcx
       488DBC2450010000     lea      rdi, [rsp+150H]
       B9CE000000           mov      ecx, 206
       33C0                 xor      rax, rax
       F3AB                 rep stosd
```
.NET Framework 32 bit zeroes out nothing:
```
011E068E 83 EC 48             sub         esp,48h  
```

**Comment by mikedn** [comment]
Tried with `unsafe`. .NET Framework x64 still can't catch up JIT32, only 64 MB/sec. CoreCLR x64 does better - 93 MB/sec. Both still generate "funny" code like:
```asm
488B8C2400040000     mov      rcx, bword ptr [rsp+400H]
48894C2440           mov      qword ptr [rsp+40H], rcx
488B4C2440           mov      rcx, qword ptr [rsp+40H]
C1E810               shr      eax, 16
0FB6C0               movzx    rax, al
4863C0               movsxd   rax, eax
```

**Comment by AndyAyersMS** [comment]
Hmm, looks like we never got to the bottom of this one. 

I'll add it to my list of things to look at.

**Comment by mikedn** [comment]
> Tried with unsafe. .NET Framework x64 still can't catch up JIT32, only 64 MB/sec. CoreCLR x64 does better - 93 MB/sec. Both still generate "funny" code like:

Not sure if I did something stupid back then or if something was improved in the JIT since then but if you use `unsafe` code to access all involved arrays then you get 160 MB/sec. So it all revolves around range checks but it's not clear how come JIT32 managed to do better here. And the interesting part is that it's not only the range checks themselves that cause problems, it's also the fact that they introduce a large amount of locals.
